compiler/src/dotty/tools/dotc/util/Signatures.scala

```diff
@@ -41,109 +43,238 @@ object Signatures {
+              (0, List(denot.asSingleDenotation))
+            else
+              (0, Nil)
+          case other => other
         (alternativeIndex, alternatives)
-
       case _ =>
         val funSymbol = fun.symbol
         val alternatives = funSymbol.owner.info.member(funSymbol.name).alternatives
         val alternativeIndex = alternatives.map(_.symbol).indexOf(funSymbol) max 0
         (alternativeIndex, alternatives)
-    }
 
-    (paramIndex, alternativeIndex, alternatives)
+    val curriedArguments = countParams(fun, alternatives(alternativeIndex))
+    val paramIndex = params.indexWhere(_.span.contains(span)) match {
+      case -1 => (params.length - 1 max 0) + curriedArguments
+      case n => n + curriedArguments
+    }
+    val alternativeSignatures = alternatives.flatMap(toApplySignature)
+    (paramIndex, alternativeIndex, alternativeSignatures)
 
+  /**
+   * Extracts call informatioin for function in unapply context.
+   *
+   * @param span The position of the cursor
+   * @param params Current function parameters
+   * @param fun Unapply function tree
+   *
+   * @return A triple containing the index of the parameter being edited, the index of the function
+   *         being called, the list of overloads of this function).
+   */
   private def unapplyCallInfo(
     span: Span,
     fun: tpd.Tree,
     patterns: List[tpd.Tree]
-  )(using Context): (Int, Int, List[SingleDenotation]) =
-    val patternPosition = patterns.indexWhere(_.span.contains(span))
-    val activeParameter = extractParamTypess(fun.tpe.finalResultType.widen).headOption.map { params =>
-      (patternPosition, patterns.length) match
-        case (-1, 0) => 0 // there are no patterns yet so it must be first one
-        case (-1, pos) => -1 // there are patterns, we must be outside range so we set no active parameter
-        case _ => (params.size - 1) min patternPosition max 0 // handle unapplySeq to always highlight Seq[A] on elements
-    }.getOrElse(-1)
+  )(using Context): (Int, Int, List[Signature]) =
+    val resultType = unapplyMethodResult(fun)
+    val isUnapplySeq = fun.denot.name == core.Names.termName("unapplySeq")
+    val denot = fun.denot
+
+    def extractParamTypess(resultType: Type): List[List[Type]] =
+      resultType match
+        // Reference to a type which is not a type class
+        case ref: TypeRef if !ref.symbol.isPrimitiveValueClass => mapOptionLessUnapply(ref, patterns.size, isUnapplySeq)
+        // Option or Some applied type. There is special syntax for multiple returned arguments: Option[TupleN]
+        // We are not intrested in it, instead we extract proper type parameters from the Option type parameter.
+        case AppliedType(TypeRef(_, cls), (appliedType @ AppliedType(tycon, args)) :: Nil)
+            if (cls == ctx.definitions.OptionClass || cls == ctx.definitions.SomeClass) =>
+          tycon match
+            case typeRef: TypeRef if ctx.definitions.isTupleClass(typeRef.symbol) => List(args)
```

Comment:
This change fixes
```scala
object OpenBrowserCommand {
  def unapply(command: String): Option[Option[Int]] = {
    Some(Some(1))
  }

  "" match {
    case OpenBrowserCommand(@@) =>
  }
}
```
It was incorrectly handled. Right now we omit nested applied types only in case of tuple class to avoid `((Int, String))`